### PR TITLE
patterns: add Xilinx boot images

### PR DESCRIPTION
**Pull request by blmaier**
* Adds pattern for Xilinx's Zynq UltraScale+ Boot Image format.
* Boot images are generated by Xilinx's `bootgen` tool.
* Spec: Xilinx UG1283 https://docs.xilinx.com/r/en-US/ug1283-bootgen-user-guide/Zynq-UltraScale-MPSoC-Boot-and-Configuration
* Add test file generated as follows
  - Build bootgen tool from https://github.com/Xilinx/bootgen
  - Create dummy data with `dd if=/dev/zero of=image.bin count=1`
  - Create bif with `echo "img: {image.bin}" >image.bif`
  - Create test bin with
    `./bootgen -arch zynqmp -image image.bif -o xilinx_bootgen.hexpat.bin`

### CODE REVIEW

**Comment by WerWolv** [comment]
Hey, thanks a lot!

Do you have any test file available (without copyrighted code in it) that you could upload as well? Then the CI can automatically run tests on it to make sure it won't break in the future

File: patterns/xilinx_bootgen.hexpat [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    struct PartitionHeader {
+        u32 encrypted_partition_data_word_length;
+        u32 unencrypted_data_word_length;
+        u32 total_partition_word_length;
+
+        NullableWordPtr<PartitionHeader, u32> next_partition_header_offset;
+
+        u64 destination_execution_address;
+        u64 destination_load_address;
+        NullableWordPtr<std::Array<u8,parent.encrypted_partition_data_word_length*4>, u32> actual_partition_word_offset;
```
Comment (blmaier) [inline patterns/xilinx_bootgen.hexpat:211]: This is an oddity I ran into with the templating system, I must refer to the array size by `parent.x` for this to work. I'm guessing the parser is evaluating the template expression from inside NullableWordPtr.

**Comment by blmaier** [comment]
> Hey, thanks a lot!
> 
> Do you have any test file available (without copyrighted code in it) that you could upload as well? Then the CI can automatically run tests on it to make sure it won't break in the future

I generated a dummy boot image with a NULL file, it has coverage of most of the pattern.

**Comment by WerWolv** [comment]
Thank you!

**Pull request merged**